### PR TITLE
feat(profiling): add type definitions

### PR DESCRIPTION
**Pull request by JonasBa**
Add type definitions for profiling (JS Self Profiling + iOS and Android profiles)

### CODE REVIEW

**Review by markstory — APPROVED** [review]

File: static/app/types/profiling.d.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    column?: number;
+    is_application?: boolean;
+    image?: string;
+    resource?: string;
+  };
+
+  type ProfileTypes =
+    | EventedProfile
+    | SampledProfile
+    | JSSelfProfiling.SpectoProfilerTrace;
```
Comment (dashed) [inline static/app/types/profiling.d.ts:46]: @JonasBa Where is `JSSelfProfiling.SpectoProfilerTrace` defined?

File: static/app/types/profiling.d.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  };
+
+  // This extends speedscope's schema - we are keeping this as is, but we are likely to diverge as we add more
+  // sentry related features to the flamegraphs. This should happen after the MVP integration
+  type Schema = {
+    name: string;
+    activeProfileIndex: number;
+    profiles: ReadonlyArray<
+      EventedProfile | SampledProfile | JSSelfProfiling.SpectoProfilerTrace
+    >;
```
Comment (dashed) [inline static/app/types/profiling.d.ts:62]:
```suggestion
    profiles: ReadonlyArray<ProfileTypes>;
```

File: static/app/types/jsSelfProfiling.d.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
+// Type definitions for https://wicg.github.io/js-self-profiling/
+namespace JSSelfProfiling {
```
Comment (evanpurkhiser) [inline static/app/types/jsSelfProfiling.d.ts:2]: We actually haven't usually used namespaces pretty much at all. But I wonder if we should start?

Good topic for the TSC

File: static/app/types/profiling.d.ts (hunk 1 above)
Comment (JonasBa) [reply static/app/types/profiling.d.ts:46]: @dashed this is a typo from my side. Should be `JSSelfProfiling.Trace`. Do you know why the CI doesnt fail here, feels like it should scream here 🤔

File: static/app/types/jsSelfProfiling.d.ts (hunk 3 above)
Comment (JonasBa) [reply static/app/types/jsSelfProfiling.d.ts:2]: @evanpurkhiser I think here it makes sense, as the terminology is often times very similar, it just varies across formats and this makes it easier to know where it belongs :)

File: static/app/types/jsSelfProfiling.d.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    name: string;
+    resourceId: number;
+    line: number;
+    column: number;
+  };
+
+  type Resource = string;
+
+  type Trace = {
+    resources: Resource[];
```
Comment (priscilawebdev) [inline static/app/types/jsSelfProfiling.d.ts:26]:
```suggestion
    resources: string[];
```

🤷‍♀️

File: static/app/types/jsSelfProfiling.d.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  };
+
+  type Frame = {
+    name: string;
+    resourceId: number;
+    line: number;
+    column: number;
+  };
+
+  type Resource = string;
```
Comment (priscilawebdev) [inline static/app/types/jsSelfProfiling.d.ts:23]:
```suggestion
```

File: static/app/types/jsSelfProfiling.d.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+  type Sample = {
+    timestamp: number;
+    stackId?: number;
+    marker?: Marker;
+  };
+
+  type Stack = {
+    parentId?: number;
+    frameId: number;
```
Comment (priscilawebdev) [inline static/app/types/jsSelfProfiling.d.ts:13]:
```suggestion
    frameId: number;
    parentId?: number;
```

File: static/app/types/profiling.d.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+  }
+
+  // Android traces follow this format
+  interface EventedProfile extends RawProfileBase {
+    events: ReadonlyArray<{at: number; frame: number; type: 'O' | 'C'}>;
+    type: 'evented';
+  }
+
+  // iOS traces follow this format
+  interface SampledProfile extends RawProfileBase {
```
Comment (priscilawebdev) [inline static/app/types/profiling.d.ts:17]: we usually only use `type` instead of `interface`

Comment (JonasBa) [reply static/app/types/profiling.d.ts:17]: @priscilawebdev 👍🏼 makes sense, but in this case (extending the base type) I would say that using interfaces is preferable, as they'll be more gracefully handled by the [compiler](https://github.com/microsoft/TypeScript-wiki/blob/main/Performance.md#preferring-interfaces-over-intersections). 

I've traced the sentry codebase to see why tsc is slow and in a lot of places, the compiler is actually struggling on really large intersection types - I'll share what I found with the team during the next TSC I think.

File: static/app/types/jsSelfProfiling.d.ts (hunk 5 above)
Comment (JonasBa) [reply static/app/types/jsSelfProfiling.d.ts:23]: Yeah, let's not complicate indeed 😅

File: static/app/types/profiling.d.ts (hunk 7 above)
Comment (priscilawebdev) [reply static/app/types/profiling.d.ts:17]: oh, that's really interesting! thanks for investigating this 🙏

Looking forward to hearing your findings at the next TSC meeting

**Review by priscilawebdev — APPROVED** [review]

File: static/app/types/profiling.d.ts (hunk 7 above)
Comment (JonasBa) [reply static/app/types/profiling.d.ts:17]: Thank you and thanks for the review! :)

**Pull request merged**